loadgen/workflow_executor.go

```diff
@@ -0,0 +1,41 @@
+package loadgen
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.temporal.io/sdk/client"
+)
+
+// WorkflowExecutor is a very dumb executor that does nothing but run the specified workflow.
+// Scenarios using it are expected to implement all their driver logic as a workflow.
+type WorkflowExecutor struct {
```

Comment:
I thought you were going to make it even more generic where a workflow would execute omes with its args as an activity if `--in-workflow` was given to the scenario or something. With this way, arguably this executor has no big value since anyone can just make an executor that makes a simple client call to start a single workflow.

Reply:
The problem with that is you don't end up getting any durability guarantees which is important. I figure as we add more CI type scenarios like this one they can abstract out the shared bits of useful workflow logic

Reply:
My fear is scenarios are now gonna have to choose whether they are wrapped as a workflow or not, which really feels more like a operator concern than a scenario author one.

Reply:
> If we agree with that, though, we can open an item in here?

I don't agree here. I think a _required_ extra layer of indirection from the caller to the scenario is a bit much. It's not whether we know Temporal will always be present, it's that a scenario is "any code" as opposed to "any workflow code". Scenario authors need a Temporal client, which is not available in a workflow. So it'll get confusing because some scenarios want workflow code and others want normal Go code.

> we should have a layer on top of every scenario that allows it to be run as a workflow.

The problem here is, like Spencer said, you can't do the things in workflow to ensure it's durable. Just a wrapper workflow that is a single max-attempt-1 activity that just runs the thing while heartbeating doesn't give you many workflow benefits.

Maybe we can have a middleground here. Maybe we leave default as it is today for most scenarios. And there is a "wrap everything in workflow" CLI option whose workflow by default is the single-max-attempt-1-activity. For scenarios that must be run as a workflow because that have in-overarching-workflow code to write in addition to normal Temporal client code, they can set a different overarching workflow type and _require_ that CLI option be given for its scenario.

Reply:
> I feel like this PR is taking the project in the wrong direction. This isn't what the original vision was.

This is I think maybe a bit not helpful, because we identified a need that we can't really meet with the existing setup, so we have to evolve to address that.

And, then yeah, we've covered why just slapping a workflow layer on top isn't really sufficient.

What I'm curious to hear is why that original vision matters, or why
> a required extra layer of indirection from the caller

Is even necessarily a bad thing. Not that I definitely disagree, it's just worth questioning. Conversely, there's clearly quite a bit of value gained by having the drivers be workflows in the form of durability. It could even work by just using embedded server specifically for the driver workflow, when a scenario isn't targeting a temporal instance (which, I still question if that's something worth caring about, we're testing temporal after all). So, what really is the "vision", and why is it important?

> there is a "wrap everything in workflow" CLI option whose workflow by default is the single-max-attempt-1-activity that just runs omes

This I think doesn't really provide much value - because the main point is that consumers who want to operate in workflow mode want to do so by using the `temporal` CLI or a client or whatever. I only made the workflow executor to just have a really easy way to run the scenario locally, and allow it to be executable using the normal `omes` CLI flags.

> For scenarios that must be run as a workflow .. they can set a different overarching workflow type and require that CLI option be given for its scenario.

This is kinda what I did, no? Except there's no flag, it just happens, that's the way this scenario runs.

Reply:
I'm not sure what needs to be durable and can't be without a workflow.

Looks like there are a couple of steps here:
1. Launch a bunch of workflows with a concurrency limit.
2. Query visibility to verify proper scenario completion.

For (1), using child workflows lets you resume from crashes without losing any progress but adds significant overhead to every start workflow call. If we were to take a different approach allowing omes to be runnable as an activity, we could use heartbeats for checkpoints without the child workflow overhead.

For (2), We could add a `validate` step that can be launched by an external coordinator for any scenario, or use the same heartbeat details to resume at that step.

I would consider not running directly as an activity, and keeping as a separate command and using the client to heartbeat and checkpoint if given an activity task token at startup.

The reason this approach takes the project away from the original vision is that one of the goals for omes was to automatically instrument workflow E2E latency and throughput, putting everything in a single workflow makes this impossible.

If you still want the child workflow approach, I would integrate it more tightly into the framework, e.g. create a `GenericChildWorkflowExecutor` and capture the required metrics.

Also your approach requires every workflow based scenario to be registered with the worker, which beats the point of having scenarios that are self contained.

Reply:
> This I think doesn't really provide much value

Only in that unfortunately some external systems (CICD) require a workflow to run anything. Other than that, I agree there's no value.

> This is kinda what I did, no? Except there's no flag, it just happens, that's the way this scenario runs.

Yup, this works for me.

My whole concern is that it's a bit confusing for scenario authors to have to concern themselves with whether where their scenario runs requires an overarching workflow. But given the lack of "any scenario can be a workflow if the caller _must_ have one", just a workflow for this one makes sense.

Reply:
@bergundy Your suggestions can work, but I have two concerns:

1. It doesn't solve the original issue, which is that the pipeline(s) want an easy way to trigger a scenario by simply starting a workflow, rather than needing a copy of the omes binary/image.
2. This is all specific to this scenario. We can do durability the hard(er) way, or we can use our own product, and the latter seems preferable.

If the only things stopping us from doing that are 
> child workflow overhead.

Well, just use a local activity to start one.
And
> automatically instrument workflow E2E latency and throughput

It seems we can also solve that by simply pulling that info from visibility and/or metrics. I'm not sure I follow why that doesn't work with a workflow driver.

Lastly, and probably most importantly, right now we have a tool which _isn't used_. These changes are motivated by _wanting to use it_. I want to be a little less concerned with philosophy and a little more concerned with actually running something and iterating on it.

Reply:
> It doesn't solve the original issue, which is that the pipeline(s) want an easy way to trigger a scenario by simply starting a workflow, rather than needing a copy of the omes binary/image.

Why can't it launch omes as a container from an activity? We're trying to come up with a standard way to launch any scenario, our pipelines should be able to run any omes scenario, not just this one.

> This is all specific to this scenario. We can do durability the hard(er) way, or we can use our own product, and the latter seems preferable. There are likely to be other concerns that your specific fixes may not address.

Is this really a requirement or are we just trying to stick to what we had? Can we come up with a solution that works for any omes scenario and worker combination?

> It seems we can also solve that by simply pulling that info from visibility and/or metrics. I'm not sure I follow why that doesn't work with a workflow driver. And also isn't actually something we really need at the moment, since we need a loadgen tool, and don't actually have a big need for a genuine benchmarking tool at the moment - not that we never will of course.

I don't see a point in a loadgen tool that doesn't do any measurements TBH, for me this is one of the top goals of this project. Sure you can find different ways to measure, but I think there's value in having the default mode measured automatically and not have to invent custom ways to measure each scenario.

> Lastly, and probably most importantly, right now we have a tool which isn't used. These changes are motivated by wanting to use it. I want to be a little less concerned with philosophy and a little more concerned with actually running something and iterating on it.

I fear we'll end up building the same tool that we're trying to replace if we're not opinionated in the way that we structure scenarios.

I think if you just want to run a single workflow, why even put it in omes?

I'm not opposed to having a workflow driven executor but I would implement it as an executor that can run arbitrary workflows just like the generic executor and it should come with OOTB support for metrics.

Reply:
Yes, it was a real requirement, from the folks who are actually going to be running this in pipelines, and they want it to be trigger-able as a workflow - and that makes good sense.

If the only thing we're "losing" is this as-yet-unused execution histogram, I'm just not seeing how that's a big issue. We are going to be tracking status using our existing metrics that Temporal emits

>not have to invent custom ways to measure each scenario.

That's exactly it, this additional histogram is a new, custom way, whereas we already have the expertise and dashboards built up around our existing metrics. We get exactly what we need with `workflow_endtoend_latency` right now.

> I would implement it as an executor that can run arbitrary workflows just like the generic executor and it should come with OOTB support for metrics.

It is? That's `workflow_executor`, there's just not much to it. I've even added the histogram metric to it, it's just maybe not particularly interesting in this case since it'll be very long times.